lib/trino-hdfs/src/test/java/io/trino/hdfs/s3/TestS3HadoopPaths.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.hdfs.s3;
+
+import io.trino.filesystem.Location;
+import org.apache.hadoop.fs.Path;
+import org.testng.annotations.Test;
+
+import java.net.URI;
+
+import static io.trino.filesystem.hdfs.HadoopPaths.hadoopPath;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class TestS3HadoopPaths
+{
+    @Test
+    public void testNonS3Unchanged()
+    {
+        assertThat(hadoopPath(Location.of("gcs://test/abc//xyz")))
+                .isEqualTo(new Path("gcs://test/abc/xyz"));
+    }
+
+    @Test
+    public void testS3NormalPath()
+    {
+        assertThat(hadoopPath(Location.of("s3://test/abc/xyz.csv")))
+                .isEqualTo(new Path("s3://test/abc/xyz.csv"))
+                .extracting(TrinoS3FileSystem::keyFromPath)
+                .isEqualTo("abc/xyz.csv");
+    }
+
+    @Test
+    public void testS3NormalPathWithInvalidUriEscape()
+    {
+        assertThat(hadoopPath(Location.of("s3://test/abc%xyz")))
+                .isEqualTo(new Path("s3://test/abc%xyz"))
+                .extracting(TrinoS3FileSystem::keyFromPath)
+                .isEqualTo("abc%xyz");
+    }
+
+    @Test
+    public void testS3NonCanonicalPath()
+    {
+        assertThat(hadoopPath(Location.of("s3://test/abc%xyz//test")))
```

Comment:
Double slash and `%` are two different problems.
The original code would pass (but incorrectly) on double slash and would fail on a %.
let's have this as two different test cases.

Reply:
No, the original code would pass with just `%`, since that would be a normalized path and thus would not use this conversion code. This test includes a `%` to ensure that we handle it in the conversion case (which the original code did not). We already have a test above for `%` without double slash.